examples/text_labels_and_annotations/date.py

```diff
@@ -36,33 +33,42 @@
 # column.
 data = cbook.get_sample_data('goog.npz', np_load=True)['price_data']
 
-fig, ax = plt.subplots()
-ax.plot('date', 'adj_close', data=data)
+fig, axs = plt.subplots(3, 1, figsize=(6.4, 7), constrained_layout=True)
+# common to all three:
+for ax in axs:
+    ax.plot('date', 'adj_close', data=data)
+    # Round to nearest years.
+    datemin = np.datetime64(data['date'][0], 'Y')
+    datemax = np.datetime64(data['date'][-1], 'Y') + np.timedelta64(1, 'Y')
+    # Minor ticks every month.
+    fmt_month = mdates.MonthLocator()
+    ax.xaxis.set_minor_locator(fmt_month)
+    fmt_half_year = mdates.MonthLocator(interval=6)
+    ax.xaxis.set_major_locator(fmt_half_year)
+
+    ax.set_xlim(datemin, datemax)
+    ax.grid(True)
+    ax.set_ylabel(r'Price [\$]')
+    # Format the coords message box, i.e. the numbers displayed as the cursor
+    # moves across the axes within the interactive GUI.
+    ax.format_ydata = lambda x: f'${x:.2f}'  # Format the price.
 
-# Major ticks every 6 months.
-fmt_half_year = mdates.MonthLocator(interval=6)
-ax.xaxis.set_major_locator(fmt_half_year)
+# different formats:
+ax = axs[0]
+ax.set_title(' DefaultFormatter', loc='left', y=0.85, fontsize='medium')
 
-# Minor ticks every month.
-fmt_month = mdates.MonthLocator()
-ax.xaxis.set_minor_locator(fmt_month)
+ax = axs[1]
+locator = ax.xaxis.get_major_locator()
+ax.xaxis.set_major_formatter(mdates.ConciseDateFormatter(locator))
+ax.set_title('  ConciseFormatter', loc='left', y=0.85, fontsize='medium')
 
+ax = axs[2]
+ax.set_title('  Manual DateFormatter', loc='left', y=0.85, fontsize='medium')
 # Text in the x axis will be displayed in 'YYYY-mm' format.
 ax.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m'))
-
-# Round to nearest years.
-datemin = np.datetime64(data['date'][0], 'Y')
-datemax = np.datetime64(data['date'][-1], 'Y') + np.timedelta64(1, 'Y')
-ax.set_xlim(datemin, datemax)
-
-# Format the coords message box, i.e. the numbers displayed as the cursor moves
-# across the axes within the interactive GUI.
-ax.format_xdata = mdates.DateFormatter('%Y-%m')
-ax.format_ydata = lambda x: f'${x:.2f}'  # Format the price.
-ax.grid(True)
-
-# Rotates and right aligns the x labels, and moves the bottom of the
-# axes up to make room for them.
-fig.autofmt_xdate()
+# Rotates and right-aligns the x labels so they don't crowd each other.
+for label in ax.get_xticklabels(which='major'):
+    label.set_ha('right')
+    label.set_rotation(30)
```

Comment:
It bugs me that we cannot do this with `tick_params`. Looping over labels feels too low-level. We might want to add label alignment options.
for now, I suggest
```suggestion
    label.set(rotation=30, horizontalalignment='right')
```

Reply:
Agreed about `tick_params`   Our tick handling needs an overhaul...